contracts/external/EstateProtocolWhitelistSTO.sol

```diff
@@ -1,145 +1,204 @@
-      return (futureBlockTimestamp, futureBlockTimestamp, pastBlockTimestamp, uint8(1));
+    modifier onlyAdmin() {
+        require(isAdmin[msg.sender], "Not an admin");
+        _;
     }
-    
-  }
 
-  function getTokenTransferStatus(address token) external view returns (bool) {
-    return tokenTransferStatus[token];
-  }
+    modifier onlyOperator() {
+        require(isOperator[msg.sender], "Not a operator");
+        _;
+    }
 
+    /**
+     * @notice Update the Merkle root for KYC verification
+     * @param root New Merkle root containing approved KYC data
+     */
+    function modifyKYCData(bytes32 root) external onlyOperator {
+        _root = root;
+        emit MerkleRootUpdated(_root);
+    }
 
+    function hasVerificationData(address investor) public view returns (bool) {
+        return verificationDataMap[investor].expiry > 0;
+    }
+
+    function addTokenLockStartTime(
+        address token,
+        uint64 startTime
+    ) external onlyAdmin {
+        tokenLockStartTime[token] = startTime;
+
+        emit TokenLockStartTimeAdded({token: token, startTime: startTime});
+    }
+
+    function modifyTokenTransferStatus(
+        address token,
+        bool status
+    ) external onlyAdmin {
+        tokenTransferStatus[token] = status;
+        emit TokenTransferStatus(token, status);
+    }
+
+    function grantAdminRole(address account) external onlyAdmin {
+        require(!isAdmin[account], "Account is already an admin");
+        isAdmin[account] = true;
+        emit AdminRoleGranted(account);
+    }
+
+    function revokeAdminRole(address account) external onlyAdmin {
+        require(isAdmin[account], "Account is not an admin");
+        isAdmin[account] = false;
+        emit AdminRoleRevoked(account);
+    }
+
+    function grantOperatorRole(address account) external onlyAdmin {
+        require(!isOperator[account], "Account is already a operator");
+        isOperator[account] = true;
+        emit OperatorRoleGranted(account);
+    }
+
+    function revokeOperatorRole(address account) external onlyAdmin {
+        require(isOperator[account], "Account is not a operator");
+        isOperator[account] = false;
+        emit OperatorRoleRevoked(account);
+    }
+
+    /**
+     * @notice Get investor in the existingInvestors `investor`.
+     */
+    function isExistingInvestor(address investor) external view returns (bool) {
+        return _existingInvestors[investor];
+    }
+
+    /**
+     * @notice Verifies if an address is whitelisted using Merkle proof
+     * @param proof Merkle proof provided by the user
+     * @param investor Address of the user to verify
+     * @param expiry Expiry timestamp from the original data
+     * @param isAccredited Accreditation status from the original data
+     * @return bool Returns true if the proof is valid
+     */
+    function verifyInvestor(
+        bytes32[] calldata proof,
+        address investor,
+        uint64 expiry,
+        bool isAccredited
```

Comment:
Can we extract the from the proof ?
`uint64 expiry,
  bool isAccredited`

Reply:
did some more research and it is not possible to extract those values 

since the proof is made up of keccak256 and so the values are hashed and are irreversible as it is a one way function, this ensures the integrity of the proof